### PR TITLE
[Train][Data] Fix double step count with TF MultiWorkerMirroredStrategy and pre-sharded dataset

### PR DESCRIPTION
**Pull request by lorenzozanee**
## Description
When using `TensorflowTrainer` with `MultiWorkerMirroredStrategy` and a pre-sharded Ray Dataset, the per-epoch step count was double the expected value. This change ensures the TF dataset is consistently marked as pre-sharded to prevent re-partitioning by the distribution strategy.

## Related issues
Fixes #61838

## Additional information
N/A

### CODE REVIEW

File: python/ray/train/_internal/data_config.py [hunk 1]
```diff
@@ -122,6 +122,12 @@ def configure(
...
             if name in datasets_to_split:
+                # Use equal=True to ensure each worker receives exactly
+                # n_rows // world_size rows. This is critical for
+                # TensorFlow's MultiWorkerMirroredStrategy where the
+                # per-epoch step count is n_rows_per_shard // batch_size.
+                # Without equal sharding, a worker could receive the full
+                # dataset and double the step count (see #61838).
                 for i, split in enumerate(
                     ds.streaming_split(
                         world_size, equal=True, locality_hints=locality_hints
```
Comment (Jokasa7) [inline python/ray/train/_internal/data_config.py:133]: [P1] This changes the shared `DataConfig` used by every `DataParallelTrainer`, not only `TensorflowTrainer`. `Dataset.streaming_split(equal=True)` explicitly drops rows that cannot be divided evenly, and `OutputSplitter` drains that truncated remainder. For example, five rows with two Torch or custom workers now silently train on four rows, whereas the previous `equal=False` path preserved all five. Please keep the common split lossless and scope equalization to the TensorFlow/MultiWorker path that needs identical step counts (or use a TensorFlow-specific data config), with a non-TensorFlow regression that sums the rows observed across workers.

**Review by Jokasa7 — COMMENTED** [review]
The TensorFlow symptom is narrow, but the `equal=True` change is made in Ray Train's shared `DataConfig`. That is the default configuration of `DataParallelTrainer`, including non-TensorFlow subclasses and direct/custom trainer use.

On this exact head, `Dataset.streaming_split` documents that `equal=True` guarantees equal row counts by dropping data when necessary, and `OutputSplitter.all_inputs_done()` explicitly drains the truncated remainder after allocating `remainder // world_size` rows per output. A five-row dataset split across two ordinary Torch or custom workers therefore changes from consuming all five rows to silently consuming four. No TensorFlow strategy or `to_tf()` call is involved in those paths, so this is a cross-backend data-loss regression introduced by a TensorFlow-specific fix. The existing automated review only discusses duplicate prefetching.

Please scope equalized splitting to the TensorFlow path that requires synchronized step counts (or use a TensorFlow-specific data configuration), keep the shared default lossless, and add a non-TensorFlow regression asserting that all input rows remain observable.
